File: src/java/org/voegtle/weatherstation/server/CacheServlet.java
package org.voegtle.weatherstation.server;

import org.json.JSONException;
import org.voegtle.weatherstation.server.persistence.CacheWeatherDTO;
import org.voegtle.weatherstation.server.util.JSONConverter;

import javax.cache.Cache;
import javax.cache.CacheException;
import javax.cache.CacheFactory;
import javax.cache.CacheManager;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class CacheServlet extends AbstractInputServlet {

  private Cache cache;

  @Override
  public void init() throws ServletException {
    super.init();
    try {
      CacheFactory cacheFactory = CacheManager.getInstance().getCacheFactory();
      Map properties = new HashMap<>();
      cache = cacheFactory.createCache(properties);
    } catch (CacheException e) {
      log.severe("CacheServlet: Could not instantiate Cache");
    }
  }

  @Override
  protected void doPost(HttpServletRequest request, HttpServletResponse resp) throws ServletException, IOException {
    try {
      String encodedWeatherData = readString(request);
      log.info("received cache object: <" + encodedWeatherData + ">");
      CacheWeatherDTO cacheWeatherDTO = new JSONConverter(locationProperties).decodeWeatherDTO(encodedWeatherData);

      cache.put(cacheWeatherDTO.getId(), cacheWeatherDTO);

      returnResult(resp, "ACK");
    } catch (JSONException e) {
      log.severe("failed to decode CacheWeatherDTO");
    }
  }


}

File: src/main/java/org/voegtle/weatherstation/server/CacheServlet.kt
package org.voegtle.weatherstation.server

import com.google.appengine.api.memcache.MemcacheService
import com.google.appengine.api.memcache.MemcacheServiceFactory
import org.json.JSONException
import org.voegtle.weatherstation.server.util.JSONConverter
import java.io.IOException
import java.util.logging.Level
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class CacheServlet : AbstractInputServlet() {

  private var cache: MemcacheService = createCache()

  @Throws(ServletException::class)
  override fun init() {
    super.init()
  }

  private fun createCache(): MemcacheService = MemcacheServiceFactory.getMemcacheService()

  @Throws(ServletException::class, IOException::class)
  override fun doPost(request: HttpServletRequest, resp: HttpServletResponse) {
    try {
      val encodedWeatherData = readString(request)
      log.info("received cache object: <$encodedWeatherData>")
      val cacheWeatherDTO = JSONConverter(locationProperties!!).decodeWeatherDTO(encodedWeatherData)

      cache.put(cacheWeatherDTO.id, cacheWeatherDTO)

      returnResult(resp, "ACK")
    } catch (e: JSONException) {
      log.log(Level.SEVERE, "failed to decode CacheWeatherDTO", e)
    }

  }


}
